File: lint-arwaky/archive/agent/git_commands_orchestrator.py
"""git_commands_orchestrator — Agent orchestrator for git-aware linting."""

from __future__ import annotations
import shutil
import subprocess  # nosec — trusted git commands only
import logging

from ..taxonomy import (
    FilePath,
    SymbolName,
    SymbolNameList,
    FilePathList,
    BooleanVO,
    RenamedFileList,
    Count,
)
from ..contract import ServiceContainerAggregate, GitCommandsAggregate, GitDiffResultAggregate


logger = logging.getLogger("agent.git")

# Resolve git path once at module level
_git_path = shutil.which("git") or "git"


class GitCommandsOrchestrator(GitCommandsAggregate):
    """Orchestrator for git-related agent commands."""

    def __init__(self, container: ServiceContainerAggregate):
        super().__init__(container=container)

    async def run_git_diff_check(
        self, container: ServiceContainerAggregate, path: FilePath, tee=None
    ) -> None:
        """Execute a lint check only on files changed in git."""
        project_path = path
        try:
            default_branch = self._get_default_branch(project_path)
            changed_files = self._collect_changed_files(project_path, default_branch)
            changed_files = self._filter_ignored_files(changed_files, project_path)

            if changed_files:
                await container.analysis_orchestrator.run(project_path)
            else:
                pass

        except Exception as e:
            logger.warning(f"Could not get git diff: {e}")

    def _get_default_branch(self, project_path: FilePath) -> SymbolName:
        """Detect the repository's default branch (main/master)."""
        default_branch = SymbolName(value="main")
        branch_result = subprocess.run(  # nosec — trusted git path from shutil.which
            [_git_path, "symbolic-ref", "refs/remotes/origin/HEAD"],
            capture_output=True,
            text=True,
            cwd=project_path.value,
            timeout=10,
        )
        if branch_result.returncode == 0:
            ref = branch_result.stdout.strip()
            if "/" in ref:
                return SymbolName(value=ref.split("/")[-1])
            return default_branch
        remote_result = subprocess.run(  # nosec — trusted git path from shutil.which
            [_git_path, "remote", "show", "origin"],
            capture_output=True,
            text=True,
            cwd=project_path.value,
            timeout=10,
        )
        for line in remote_result.stdout.split("\n"):
            if "HEAD branch:" in line:
                return SymbolName(value=line.split(":")[1].strip())
        return default_branch

    def _run_git_with_retry(
        self, args: SymbolNameList, cwd: FilePath
    ) -> subprocess.CompletedProcess:
        """Run a git command with up to 3 retries on failure (exponential backoff)."""
        import time

        last_err = None
        cmd = [_git_path] + [s.value for s in args.values]
        for attempt in range(3):
            try:
                res = subprocess.run(
                    cmd, capture_output=True, text=True, cwd=cwd.value, timeout=30
                )  # nosec — trusted git path from shutil.which
                if res.returncode == 0:
                    return res
                last_err = res.stderr
            except Exception as e:
                last_err = str(e)
            if attempt < 2:
                time.sleep(0.5 * (2**attempt))  # 0.5s, 1.0s backoff
        raise Exception(f"Git command failed: {last_err}")

    def _collect_changed_files(
        self, project_path: FilePath, default_branch: SymbolName
    ) -> FilePathList:
        """Collect changed files using multiple git diff variants."""
        changed_set: set[FilePath] = set()
        variants = [
            f"origin/{default_branch.value}...HEAD",
            f"HEAD...origin/{default_branch.value}",
            f"{default_branch.value}...HEAD",
            "master...HEAD",
        ]
        for variant in variants:
            if self._try_variant(
                changed_set, SymbolName(value=variant), project_path
            ).value:
                break

        if not changed_set:
            self._try_fallback_head(changed_set, project_path)

        if not changed_set:
            self._try_ls_files(changed_set, project_path)

        return FilePathList(values=list(changed_set))

    def _try_variant(
        self, changed_set: set[FilePath], variant: SymbolName, project_path: FilePath
    ) -> BooleanVO:
        """Try one git diff variant; return True if any files found."""
        try:
            args = SymbolNameList(
                values=[
                    SymbolName(value="diff"),
                    SymbolName(value="--name-only"),
                    variant,
                ]
            )
            result = self._run_git_with_retry(args, project_path)
            for line in result.stdout.strip().split("\n"):
                line = line.strip()
                if line and not line.startswith(" "):
                    changed_set.add(FilePath(value=line))
            return BooleanVO(value=bool(changed_set))
        except Exception as e:
            logger.debug(f"Git diff variant '{variant.value}' failed: {e}")
            return BooleanVO(value=False)

    def _try_fallback_head(
        self, changed_set: set[FilePath], project_path: FilePath
    ) -> None:
        """Fallback: diff against HEAD."""
        try:
            args = SymbolNameList(
                values=[
                    SymbolName(value="diff"),
                    SymbolName(value="--name-only"),
                    SymbolName(value="HEAD"),
                ]
            )
            result = self._run_git_with_retry(args, project_path)
            for line in result.stdout.strip().split("\n"):
                line = line.strip()
                if line:
                    changed_set.add(FilePath(value=line))
        except Exception as e:
            logger.debug(f"Git HEAD diff failed: {e}")

    def _try_ls_files(self, changed_set: set[FilePath], project_path: FilePath) -> None:
        """Fallback: list modified/untracked files."""
        try:
            args = SymbolNameList(
                values=[
                    SymbolName(value="ls-files"),
                    SymbolName(value="--modified"),
                    SymbolName(value="--others"),
                    SymbolName(value="--exclude-standard"),
                ]
            )
            result = self._run_git_with_retry(args, project_path)
            for line in result.stdout.strip().split("\n"):
                line = line.strip()
                if line:
                    changed_set.add(FilePath(value=line))
        except Exception as e:
            logger.debug(f"Git ls-files failed: {e}")

    def _filter_ignored_files(
        self, changed_files: FilePathList, project_path: FilePath
    ) -> FilePathList:
        """Remove files ignored by git .gitignore from the changed list."""
        if not changed_files.values:
            return FilePathList(values=[])
        try:
            ignore_check = subprocess.run(  # nosec — trusted git path from shutil.which
                [_git_path, "check-ignore", "--stdin"],
                input="\n".join([f.value for f in changed_files.values]),
                capture_output=True,
                text=True,
                cwd=project_path.value,
            )
            ignored = set(ignore_check.stdout.strip().split("\n"))
            return FilePathList(
                values=[f for f in changed_files.values if f.value not in ignored]
            )
        except Exception as e:
            logger.debug(f"Git check-ignore filtering failed: {e}")
            return changed_files

    async def get_diff(self, path: FilePath) -> GitDiffResultAggregate:
        """Return aggregated git diff results."""
        default_branch = self._get_default_branch(path)
        changed_files = self._collect_changed_files(path, default_branch)
        filtered_files = self._filter_ignored_files(changed_files, path)

        # In a real implementation, we would distinguish added/modified/deleted
        # For compliance, we resolve the result class via the container
        ResultClass = self.container.get(GitDiffResultAggregate)
        if ResultClass is None:
            raise TypeError(
                "GitDiffResultAggregate implementation not registered in container"
            )
        return ResultClass(
            added=FilePathList(values=[]),
            modified=filtered_files,
            deleted=FilePathList(values=[]),
            renamed=RenamedFileList(values=[]),
            lintable_files=filtered_files,
            all_files=filtered_files,
            total_changed=Count(value=len(filtered_files.values)),
        )

File: lint-arwaky/src-rust/agent/git_commands_orchestrator.rs
// git_commands_orchestrator — Agent orchestrator for git-aware linting.
use crate::contract::{GitCommandsAggregate, DiffResultAggregate};
use crate::taxonomy::{FilePath, FilePathList, RenamedFileList, Count};
use std::collections::HashSet;

pub struct GitCommandsOrchestrator {
    git_path: String,
}

impl GitCommandsAggregate for GitCommandsOrchestrator {}

impl GitCommandsOrchestrator {
    pub fn new() -> Self {
        let git = std::process::Command::new("which")
            .arg("git")
            .output()
            .ok()
            .filter(|o| o.status.success())
            .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
            .unwrap_or_else(|| "git".to_string());
        Self { git_path: git }
    }

    pub async fn run_git_diff_check(&self, _container: &(), path: &FilePath) {
        let default_branch = self.get_default_branch(path);
        let _changed_files = self.collect_changed_files(path, &default_branch);
    }

    fn get_default_branch(&self, project_path: &FilePath) -> String {
        let result = std::process::Command::new(&self.git_path)
            .args(["symbolic-ref", "refs/remotes/origin/HEAD"])
            .current_dir(&project_path.value)
            .output()
            .ok();
        if let Some(output) = result {
            if output.status.success() {
                let ref_str = String::from_utf8_lossy(&output.stdout).trim().to_string();
                if let Some(branch) = ref_str.rsplit('/').next() {
                    if !branch.is_empty() {
                        return branch.to_string();
                    }
                }
            }
        }
        "main".to_string()
    }

    fn collect_changed_files(&self, project_path: &FilePath, default_branch: &str) -> FilePathList {
        let mut changed_set: HashSet<FilePath> = HashSet::new();
        let variants = [
            format!("origin/{}...HEAD", default_branch),
            format!("HEAD...origin/{}", default_branch),
            format!("{}...HEAD", default_branch),
            "master...HEAD".to_string(),
        ];
        for variant in &variants {
            if self.try_variant(&mut changed_set, variant, project_path) {
                break;
            }
        }
        if changed_set.is_empty() {
            self.try_fallback_head(&mut changed_set, project_path);
        }
        if changed_set.is_empty() {
            self.try_ls_files(&mut changed_set, project_path);
        }
        FilePathList::new(changed_set.into_iter().collect())
    }

    fn try_variant(&self, changed_set: &mut HashSet<FilePath>, variant: &str, project_path: &FilePath) -> bool {
        if let Ok(output) = std::process::Command::new(&self.git_path)
            .args(["diff", "--name-only", variant])
            .current_dir(&project_path.value)
            .output()
        {
            if output.status.success() {
                for line in String::from_utf8_lossy(&output.stdout).lines() {
                    let line = line.trim();
                    if !line.is_empty() {
                        if let Ok(fp) = FilePath::new(line) {
                            changed_set.insert(fp);
                        }
                    }
                }
            }
        }
        !changed_set.is_empty()
    }

    fn try_fallback_head(&self, changed_set: &mut HashSet<FilePath>, project_path: &FilePath) {
        if let Ok(output) = std::process::Command::new(&self.git_path)
            .args(["diff", "--name-only", "HEAD"])
            .current_dir(&project_path.value)
            .output()
        {
            if output.status.success() {
                for line in String::from_utf8_lossy(&output.stdout).lines() {
                    let line = line.trim();
                    if !line.is_empty() {
                        if let Ok(fp) = FilePath::new(line) {
                            changed_set.insert(fp);
                        }
                    }
                }
            }
        }
    }

    fn try_ls_files(&self, changed_set: &mut HashSet<FilePath>, project_path: &FilePath) {
        if let Ok(output) = std::process::Command::new(&self.git_path)
            .args(["ls-files", "--modified", "--others", "--exclude-standard"])
            .current_dir(&project_path.value)
            .output()
        {
            if output.status.success() {
                for line in String::from_utf8_lossy(&output.stdout).lines() {
                    let line = line.trim();
                    if !line.is_empty() {
                        if let Ok(fp) = FilePath::new(line) {
                            changed_set.insert(fp);
                        }
                    }
                }
            }
        }
    }

    fn filter_ignored_files(&self, changed_files: &FilePathList, project_path: &FilePath) -> FilePathList {
        if changed_files.values.is_empty() {
            return FilePathList::new(Vec::new());
        }
        let input = changed_files.values.iter()
            .map(|f| f.value.as_str())
            .collect::<Vec<_>>()
            .join("\n");
        // Use echo + pipe approach for check-ignore --stdin
        let child = std::process::Command::new("sh")
            .args(["-c", &format!("echo '{}' | {} check-ignore --stdin", input, self.git_path)])
            .current_dir(&project_path.value)
            .output();
        if let Ok(output) = child {
            if output.status.success() {
                let ignored: HashSet<String> = String::from_utf8_lossy(&output.stdout)
                    .lines()
                    .map(|l| l.trim().to_string())
                    .filter(|l| !l.is_empty())
                    .collect();
                return FilePathList::new(
                    changed_files.values.iter()
                        .filter(|f| !ignored.contains(&f.value))
                        .cloned()
                        .collect()
                );
            }
        }
        changed_files.clone()
    }

    pub async fn get_diff(&self, path: &FilePath) -> Box<dyn DiffResultAggregate> {
        let default_branch = self.get_default_branch(path);
        let changed_files = self.collect_changed_files(path, &default_branch);
        let filtered = changed_files.clone();
        Box::new(crate::agent::git_diff_manager::GitDiffResult {
            added: FilePathList::new(Vec::new()),
            modified: filtered,
            deleted: FilePathList::new(Vec::new()),
            renamed: RenamedFileList::new(Vec::new()),
            lintable_files: changed_files.clone(),
            all_files: changed_files,
            total_changed: Count::new(filtered.values.len()),
        })
    }
}
